lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java

```diff
@@ -0,0 +1,367 @@
+                previousPageLastEvaluatedHashUserId == null
+                        ? null
+                        : Map.of(
+                                HASH_USER_ID,
+                                AttributeValue.builder()
+                                        .s(previousPageLastEvaluatedHashUserId)
+                                        .build());
+        var forkJoinPool =
+                forkJoinPoolFactory.getForkJoinPool(
+                        request.parallelism() == null
+                                ? DEFAULT_PARALLELISM
+                                : request.parallelism());
+
+        LOGGER.info(
+                LogHelper.buildLogMessage("Starting bulk migration")
+                        .with(LOG_BATCH_ID.getFieldName(), batchId)
+                        .with("batchSize", request.batch().size())
+                        .with("parallelism", forkJoinPool.getParallelism())
+                        .with("exclusiveStartKey", exclusiveStartKey));
+
+        try {
+            for (var page :
+                    reportUserIdentityScanDynamoDataStore.scan(
+                            exclusiveStartKey,
+                            request.pageSize(),
+                            HASH_USER_ID,
+                            USER_ID,
+                            IDENTITY)) {
+
+                var pageSummary =
+                        new PageSummary(previousPageLastEvaluatedHashUserId, page.count());
+
+                LOGGER.info(
+                        LogHelper.buildLogMessage("Processing page")
+                                .with(PAGE_EXCLUSIVE_START_KEY, pageSummary.getExclusiveStartKey())
+                                .with(PAGE_ITEM_COUNT, pageSummary.getCount()));
+
+                // Updating for next loop while we still have access to the page
+                previousPageLastEvaluatedHashUserId = getPageLastEvaluatedHashUserId(page);
+
+                submitPageToThreadPool(forkJoinPool, page, pageSummary, batchId);
+
+                report.addPageSummary(pageSummary);
+
+                if (context.getRemainingTimeInMillis() <= ONE_MINUTE_IN_MS) {
+                    LOGGER.warn(
+                            LogHelper.buildLogMessage(
+                                            "Lambda close to timeout - stopping execution")
+                                    .with(LOG_BATCH_ID.getFieldName(), batchId)
+                                    .with(
+                                            PAGE_EXCLUSIVE_START_KEY,
+                                            pageSummary.getExclusiveStartKey())
+                                    .with(PAGE_ITEM_COUNT, pageSummary.getCount()));
+                    report.finalise(
+                            "Lambda close to timeout", getPageLastEvaluatedHashUserId(page));
+                    throw new StopBeforeLambdaTimeoutException();
+                }
+
+                if (report.getTotalEvaluated() >= request.batch().size()) {
```

Comment:
I did consider this and thought that it probably doesn't matter all that much, we won't be orders of magnitude out. I also thought it's proobably very likely that the page and batch sizes will line up, unless we decide to do something a bit odd.

Reply:
Yeah, I think we can make it work.

(I'd just like to avoid having to explain 'we said batch 1 would be X users, but it actually ended up being Y', if possible!)